src/net/socket.rs

```diff
@@ -159,70 +128,31 @@ impl Socket {
 
     // Receives a single message from the socket. On success, returns the packet containing origin and data.
     fn recv_from(&mut self) -> Result<Option<Packet>> {
-        let (recv_len, address) = self.socket.recv_from(&mut self.recv_buffer)?;
-        if recv_len == 0 {
-            return Err(ErrorKind::ReceivedDataToShort)?;
+        match self.socket.recv_from(&mut self.recv_buffer) {
+            Ok((recv_len, address)) => {
+                if recv_len == 0 {
+                    return Err(ErrorKind::ReceivedDataToShort)?;
+                }
+                let received_payload = &self.recv_buffer[..recv_len];
+                let connection = self
+                    .connections
+                    .get_or_insert_connection(address, &self.config);
+                connection.process_incoming(received_payload)
+            }
+            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                error!("Encountered a WouldBlock: {:?}", e);
```

Comment:
Why don't we return the error here? As said in a previous discussion about this line, our error type already supports mapping an io error to our own error.